File: examples/pizzas_static/lib/pizzas_domain/pizza/topping.rb
module PizzasDomain
  class Pizza
    class Topping
      attr_reader :name, :amount

      def initialize(name:, amount:)
        @name = name
        @amount = amount
        check_invariants!
        freeze
      end

      def ==(other)
        other.is_a?(self.class) &&
          name == other.name &&
          amount == other.amount
      end
      alias eql? ==

      def hash
        [self.class, name, amount].hash
      end

      private

      def check_invariants!
        raise InvariantError, "amount must be positive" unless instance_eval(&proc { true })
      end
    end
  end
end

File: examples/pizzas_static_go/domain/topping.go
package domain

import "fmt"

type Topping struct {
	Name string `json:"name"`
	Amount int64 `json:"amount"`
}

func NewTopping(name string, amount int64) (Topping, error) {
	v := Topping{
		Name: name,
		Amount: amount,
	}
	// amount must be positive
	if v.Amount <= 0 {
		return Topping{}, fmt.Errorf("amount must be positive")
	}
	return v, nil
}
